### PR TITLE
Added OpenVX based processing to LUT

### PR DESCRIPTION
**Pull request by terfendail**
### This pullrequest changes
Implemented LUT image processing via OpenVX TableLookup immediate mode call

### CODE REVIEW

File: 3rdparty/openvx/include/ivx.hpp [hunk 1]
```diff
@@ -1754,6 +1763,125 @@ class Matrix : public RefWrapper<vx_matrix>
...
+#ifdef VX_VERSION_1_1
+        IVX_CHECK_STATUS(vxCopyLUT(ref, data, usage, memType));
+#else
+        IVX_CHECK_STATUS(vxAccessLUT(ref, const_cast<void**>(&data), usage));
+        IVX_CHECK_STATUS(vxCommitLUT(ref, data));
+#endif
+    }
+
+#ifdef IVX_USE_OPENCV
+    void createMat(cv::Mat& m)
```
Comment (apavlenko) [inline 3rdparty/openvx/include/ivx.hpp]: I see no much value in this method...

File: 3rdparty/openvx/include/ivx.hpp [hunk 2]
```diff
@@ -1754,6 +1763,125 @@ class Matrix : public RefWrapper<vx_matrix>
...
+    {
+        m.create(1, (int)count(), enumToCVType(dataType()));
+    }
+
+    void copyTo(cv::Mat& m)
+    {
+        createMat(m);
```
Comment (apavlenko) [inline 3rdparty/openvx/include/ivx.hpp]: I suggest making allocation of an empty Mat only and shape checking for non-empty.

File: modules/core/src/convert.cpp [hunk 3]
```diff
@@ -5287,6 +5292,43 @@ static bool ocl_LUT(InputArray _src, InputArray _lut, OutputArray _dst)
...
+{
+    if (src.type() != CV_8UC1 || dst.type() != src.type() || _lut.type() != src.type() || !_lut.isContinuous())
+        return false;
+
+    try
+    {
+        ivx::Context ctx = ivx::Context::create();
+
+        ivx::Image
+            ia = ivx::Image::createFromHandle(ctx, VX_DF_IMAGE_U8,
```
Comment (apavlenko) [inline modules/core/src/convert.cpp]: consider the shorter code:
```cpp
ia = ivx::Image::createFromHandle(
    ctx,
    ivx::Image::matTypeToFormat(src.type()),
    ivx::Image::createAddressing(src),
    src.data );
```

And I'm actually thinking about adding `ivx::Image::createFromHandle(context, mat);` to ivx.hpp, but this require saving that Mat reference inside the created Image...

File: modules/core/src/convert.cpp [hunk 4]
```diff
@@ -5287,6 +5292,43 @@ static bool ocl_LUT(InputArray _src, InputArray _lut, OutputArray _dst)
...
+    catch (ivx::RuntimeError & e)
+    {
+        CV_Error(CV_StsInternal, e.what());
+        return false;
+    }
+    catch (ivx::WrapperError & e)
+    {
+        CV_Error(CV_StsInternal, e.what());
+        return false;
+    }
```
Comment (apavlenko) [inline modules/core/src/convert.cpp]: Looks like we need a base class `ivx::Error` to prevent duplicated code for both types of ivx exceptions...

Comment (terfendail) [inline modules/core/src/convert.cpp]: `ivx::Error` looks useful but I think such refactoring could be postponed to the moment when most of kernel wrappers be merged

Comment (terfendail) [inline modules/core/src/convert.cpp]: `ivx::Error` looks useful, but I think it would be better to refactor it later

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**